src/plotting/plot_utils.py

```diff
@@ -96,4 +102,194 @@ def read_satinfo_files(satinfo_db_root,satinfo_string):
+      ozinfo.loc[len(ozinfo.index)] = [date(2100,1,1), False, np.nan]
+    #convert logical to floats with nans for plotting
+    ozinfo['status_nan'] = ozinfo.status.astype('int')
+    ozinfo['status_nan'].replace(0, np.nan, inplace=True)
+    #make sure the end of the series is in the future
+    ozinfo.loc[len(ozinfo.index)]=[date(2100,1,1), ozinfo.status.iat[-1], ozinfo.status_nan.iat[-1]]    
+    ozinfo.datetime = pandas.to_datetime(ozinfo.datetime)
+    return ozinfo
+
+def get_distinct_bufr():
+    session = itf.Session()
+    # Subquery to get the most recent inserted_at for each combination of other columns
+    subquery = session.query(
+        omnb.obs_id,
+        omnb.sat_id,
+        omnb.sat_id_name,
+        omnb.obs_count,
+        omnb.sat_inst_id,
+        omnb.sat_inst_desc,
+        omnb.filename,
+        omnb.file_size,
+        omnb.obs_day,
+        func.max(omnb.inserted_at).label('max_inserted_at')
+    ).group_by(
+        omnb.obs_id,
+        omnb.sat_id,
+        omnb.sat_id_name,
+        omnb.obs_count,
+        omnb.sat_inst_id,
+        omnb.sat_inst_desc,
+        omnb.filename,
+        omnb.file_size,
+        omnb.obs_day
+    ).subquery()
+
+    # Join the subquery with the main table to get the full records
+    query = session.query(omnb.obs_id, omnb.filename, omnb.sat_id, omnb.sat_id_name, omnb.obs_count, omnb.obs_day, omnb.file_size, oi.parent_dir, oi.s3_bucket).join(
+        subquery,
+        (omnb.obs_id == subquery.c.obs_id) &
+        (omnb.sat_id == subquery.c.sat_id) &
+        (omnb.obs_count == subquery.c.obs_count) &
+        (omnb.sat_inst_id == subquery.c.sat_inst_id) &
+        (omnb.filename == subquery.c.filename) &
+        (omnb.file_size == subquery.c.file_size) &
+        (omnb.obs_day == subquery.c.obs_day) &
+        (omnb.inserted_at == subquery.c.max_inserted_at)
+    ).join(
+        oi,
+        omnb.obs_id == oi.obs_id
+    ).filter(
+        oi.s3_bucket == 'noaa-reanalyses-pds'
+    )
+
+    # Execute the query
+    results = query.all()
+
+    # Convert results to a list of dictionaries
+    result_dicts = [
+        {
+            'obs_id': result.obs_id,
+            'filename': result.filename,
+            'sat_id': result.sat_id,
+            'sat_id_name': result.sat_id_name,
+            'obs_count': result.obs_count,
+            'obs_day': result.obs_day,
+            'file_size': result.file_size,
+            'parent_dir': result.parent_dir,
+            's3_bucket': result.s3_bucket
+        }
+        for result in results
+    ]
+
+    # Convert the list of dictionaries to a pandas DataFrame
+    df = pandas.DataFrame(result_dicts)
+
+    # Close the session
+    session.close()
+
+    return df
+
+def get_distinct_prepbufr():
+    session = itf.Session()
+    # Subquery to get the most recent inserted_at for each combination of other columns
+    subquery = session.query(
+        omnp.obs_id,
+        omnp.cmd_str,
```

Comment:
I think this subquery needs to be trimmed down to match the main prep buffer join as well. Everything else looks great.

Reply:
my concern with trimming this part down is that it's doing something different than below. This is to reduce duplicates, the below is about what data is needed for plotting.

Reply:
I see. I think if we don't select the command string and qm4+ columns, wouldn't we potentially eliminate even more duplicates with the group by? Though that would only work if we don't care about the data in those columns. This would also make it more similar to the buffer subquery.

Reply:
I don't think the group by is going to make a difference because it will need to match what's called in the query.

Removing command string is a good idea and I think it'll be fine to remove the qm markers since the tot and qm0thru3 should be enough if there's a change in the other qms, the only duplicates we wouldn't necessary catch is file changes which should be less frequent. I'll take care of that change and test it

Reply:
oh just for your future information, it's not really comparable to the bufr because there's nothing besides obs_count for bufr files whereas the prepbufr has quality marker breakdowns.

Reply:
updated and tested